<h1>Описание проекта</h1>

Для формулирования ценностного предложения для нового продукта необходимо проанализировать базу данных с информацией о книгах, издательствах, авторах и пользовательских обзорах

<h1>Описание данных</h1>

Таблица books

Содержит данные о книгах:

book_id — идентификатор книги;

author_id — идентификатор автора;

title — название книги;

num_pages — количество страниц;

publication_date — дата публикации книги;

publisher_id — идентификатор издателя.

Таблица authors

Содержит данные об авторах:

author_id — идентификатор автора;

author — имя автора.

Таблица publishers

Содержит данные об издательствах:

publisher_id — идентификатор издательства;

publisher — название издательства;

Таблица ratings

Содержит данные о пользовательских оценках книг:

rating_id — идентификатор оценки;

book_id — идентификатор книги;

username — имя пользователя, оставившего оценку;

rating — оценка книги.

Таблица reviews

Содержит данные о пользовательских обзорах:

review_id — идентификатор обзора;

book_id — идентификатор книги;

username — имя автора обзора;

text — текст обзора.

<h1>Загрузка данных</h1>

Подключаемся к базе данных

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Создаем функцию запроса к таблице

In [3]:
def query(text, data, index):
    query = text
    data = pd.io.sql.read_sql(text, con=engine, index_col = index)
    return data

Создаем датафреймы

In [4]:
books = pd.DataFrame()
authors = pd.DataFrame()
publishers = pd.DataFrame()
ratings = pd.DataFrame()
reviews = pd.DataFrame()

Создаем запросы к каждой таблице

In [5]:
query(''' SELECT * FROM books''', books, 'book_id')

author_id                                              title  \
book_id                                                                 
1              546                                       'Salem's Lot   
2              465                 1 000 Places to See Before You Die   
3              407  13 Little Blue Envelopes (Little Blue Envelope...   
4               82  1491: New Revelations of the Americas Before C...   
5              125                                               1776   
...            ...                                                ...   
996            571           Wyrd Sisters (Discworld  #6; Witches #2)   
997            454                        Xenocide (Ender's Saga  #3)   
998            201                                    Year of Wonders   
999             94                        You Suck (A Love Story  #2)   
1000           509  Zen and the Art of Motorcycle Maintenance: An ...   

         num_pages publication_date  publisher_id  
book_id                                            
1              594       2005-11-01            93  
2              992       2003-05-22           336  
3              322       2010-12-21           135  
4              541       2006-10-10           309  
5              386       2006-07-04           268  
...            ...              ...           ...  
996            265       2001-02-06           147  
997            592       1996-07-15           297  
998            358       2002-04-30           212  
999            328       2007-01-16           331  
1000           540       2006-04-25           143  

[1000 rows x 5 columns]

In [6]:
query(''' SELECT * FROM authors''', authors, 'author_id')

author
author_id                                      
1                                    A.S. Byatt
2                Aesop/Laura Harris/Laura Gibbs
3                               Agatha Christie
4                                 Alan Brennert
5                      Alan Moore/David   Lloyd
...                                         ...
632               William Strunk Jr./E.B. White
633                                 Zadie Smith
634                       Zilpha Keatley Snyder
635                          Zora Neale Hurston
636        Åsne Seierstad/Ingrid Christopherson

[636 rows x 1 columns]

In [7]:
query(''' SELECT * FROM publishers''', publishers, 'publisher_id')

publisher
publisher_id                                   
1                                           Ace
2                                      Ace Book
3                                     Ace Books
4                                 Ace Hardcover
5             Addison Wesley Publishing Company
...                                         ...
336                  Workman Publishing Company
337                                  Wyatt Book
338                       Yale University Press
339                                    Yearling
340                              Yearling Books

[340 rows x 1 columns]

In [8]:
query(''' SELECT * FROM ratings''', ratings, 'rating_id')

book_id        username  rating
rating_id                                 
1                1      ryanfranco       4
2                1   grantpatricia       2
3                1    brandtandrea       5
4                2        lorichen       3
5                2     mariokeller       2
...            ...             ...     ...
6452          1000  carolrodriguez       4
6453          1000         wendy18       4
6454          1000      jarvispaul       5
6455          1000           zross       2
6456          1000         fharris       5

[6456 rows x 3 columns]

In [9]:
query(''' SELECT * FROM reviews''', reviews, 'review_id')

book_id        username  \
review_id                            
1                1    brandtandrea   
2                1      ryanfranco   
3                2        lorichen   
4                3   johnsonamanda   
5                3     scotttamara   
...            ...             ...   
2789           999      martinadam   
2790          1000         wknight   
2791          1000  carolrodriguez   
2792          1000         wendy18   
2793          1000      jarvispaul   

                                                        text  
review_id                                                     
1          Mention society tell send professor analysis. ...  
2          Foot glass pretty audience hit themselves. Amo...  
3          Listen treat keep worry. Miss husband tax but ...  
4          Finally month interesting blue could nature cu...  
5          Nation purpose heavy give wait song will. List...  
...                                                      ...  
2789       Later hospital turn easy community. Fact same ...  
2790        Change lose answer close pressure. Spend so now.  
2791       Authority go who television entire hair guy po...  
2792       Or western offer wonder ask. More hear phone f...  
2793       Republican staff bit eat material measure plan...  

[2793 rows x 3 columns]

<h1>SQL запросы</h1>

Создаем функцию запроса

In [10]:
def sql_query(text):
    query = text
    return pd.io.sql.read_sql(text, con=engine)

1. Посчитайте, сколько книг вышло после 1 января 2000 года

In [11]:
sql_query(''' SELECT COUNT(book_id) 
FROM books 
WHERE publication_date >= '2000-01-01' 
''')

count
0    821

821 книга вышла после 1 января 2000 года

2. Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [12]:
sql_query(''' WITH
rev AS (
    SELECT b.book_id, COUNT(review_id) as review_count
    FROM books b
    JOIN reviews re
    ON b.book_id = re.book_id
    GROUP BY b.book_id
    ),
rat AS (
    SELECT b.book_id, AVG(rating) as avg_rating
    FROM books b
    JOIN ratings ra
    ON b.book_id = ra.book_id
    GROUP BY b.book_id
    )

SELECT rev.book_id, review_count, avg_rating
FROM rev
JOIN rat
ON rev.book_id = rat.book_id
''')

book_id  review_count  avg_rating
0          1             2    3.666667
1          2             1    2.500000
2          3             3    4.666667
3          4             2    4.500000
4          5             4    4.000000
..       ...           ...         ...
989      996             3    3.666667
990      997             3    3.400000
991      998             4    3.200000
992      999             2    4.500000
993     1000             4    3.833333

[994 rows x 3 columns]

Результаты приведены в таблице выше

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [13]:
sql_query(''' SELECT p.publisher_id, publisher, COUNT(book_id) as book_count
FROM publishers p
JOIN books b ON p.publisher_id = b.publisher_id
WHERE num_pages > 50
GROUP BY p.publisher_id, publisher
ORDER BY book_count DESC
LIMIT 1
 ''')

publisher_id      publisher  book_count
0           212  Penguin Books          42

Издательство Penguin Books (id 212) выпустило 42 книги толще 50 страниц, что является наибольшим результатом среди представлененых в базе издательств

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [34]:
sql_query(''' WITH
rat as (
    SELECT b.book_id, author_id
    FROM books b
    JOIN ratings ra
    ON b.book_id = ra.book_id
    GROUP BY b.book_id
    HAVING COUNT(rating_id) >= 50
    )

SELECT a.author_id, author, AVG(rating) as avg_rating
FROM authors a
JOIN rat
ON a.author_id = rat.author_id
JOIN ratings r
ON rat.book_id = r.book_id
GROUP BY a.author_id, author
ORDER BY AVG(rating) desc
LIMIT 1
 ''')

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.287097

Джоан Роулинг является автором с наивысшей средней оценкой, учитывая только книги с 50 и более оценками

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [ ]:
sql_query('''
WITH rating_count AS 
(
    SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(rating_id) > 50
)
 
SELECT AVG(review_count)
FROM 
(
    SELECT COUNT(review_id) as review_count
    FROM reviews re
    JOIN rating_count rc ON re.username = rc.username
    GROUP BY rc.username
) AS review_count

''')

Пользователи, которые поставили больше 50 оценок, в среднем обозревают 24.3 книги

В данной работе была исследована база данных с информацией о книгах, авторах, издательствах и пользовательских обзорах. В базе в основном представлены современные книги (примерно 80% книг выпущены после 2000 года). Наиболее успешным автором является Джоан Роулинг. Думаю многие сталкивались с ее полюбившейся многим серией о Гарри Поттере, поэтому не удивительно, что у этого автора наибольшая средняя оценка среди авторов книг с множеством оценок. Самым активным издателем за эти годы стало издательство Penguin Books. Это издательство является одним из самых известных во всем мире. Именно это издательство, как считают многие, привнесло чтение в массы. Наиболее важным критерием успешности книги является оценка читателей. Но не менее важны комментарии / обзоры. Активные пользователи (у кого больше 50 оценок) в среднем обозревают половину оцененных ими книг. Обзоры пользователей совместно с оценками помогут правильно выбрать направление развития библиотеки, выбрать какие книги больше нравятся читателям и добавлять похожие.